<h1 style="text-align: center;">Desafío - Funciones y probabilidades</h1>

- Para realizar este desafío debes haber estudiado previamente todo el material disponible correspondiente a la unidad.
- Una vez terminado el desafío, comprime la carpeta que contiene el desarrollo de los requerimientos solicitados y sube el .zip en el LMS.
- Para la realización del desafío necesitarás apoyarte del archivo `Apoyo Desafío - Calculando probabilidades.zip`, disponible en la <a href="https://empieza.desafiolatam.com/cohorts/2091/sections/78268/activities/119096/items/610788">plataforma</a>.
- Desarrollo desafío:
    - El desafío se debe desarrollar de manera Individual.
    - Puntaje total: 10 puntos.

## Desafío 1: Generación de funciones (1 punto)
- Genere funciones para calcular la media y varianza de un vector. Debe cumplir con los siguientes requisitos:
    - Ambas funciones deben ingresar un argumento x correspondiente al vector.
    - Las funciones deben contener docstrings con la documentación asociada a la variable y retornar un valor utilizando `return`.
- La función de la varianza se debe llamar a la función de la media.
- Utilice las funciones para reportar la información sobre `goles_favor`, `goles_contra` y `puntos`.

In [45]:
# Utiliza este bloque para definir las funciones solicitadas

#Un vector es una lista de datos

a = [1,2,3,4,5,6,7,8,9]

def media(x):
    """
    Definición: Esta función entrega el promedio de n numeros.
    
    Parametros de ingreso: n numeros del tipo int o float, los cuales se suman en su totalidad y se divide por la cantidad
    de numeros entregados.
    
    Retorno: Devuelve el resultado de la suma de todos los numeros dividido por el largo o la cantidad de numeros
    
    """
    return sum(x) / len(x)

media(a)


5.0

In [46]:
def varianza(x):
    """
    Definición: Esta función entrega la varianza de n numeros.
    
    Parametros de ingreso: n numeros del tipo int o float, el cual es una sumatoria entre cada numero - el promedio 
    (cabe destacar que para este numero invoca la funcion media declara más arriba) y esté es elevado al cuadrado para finalmente ser dividido por la cantidad de numeros.
    
    Retorno: Devuelve un float que es la varianza de la cantidad de numeros que entregamos como parametros.
    """
    var = 0
    
    for i in x:
        var = var + (((i-media(x))**2)/len(x))
    
    return var

varianza(a)

6.666666666666667

In [47]:
# Utiliza este bloque para reportar la información sobre goles_favor, goles_contra y puntos

import pandas as pd
import numpy as np
df = pd.read_csv("worldcup2014.csv")
df.head(5)


team     continent group  group_pos  cantidad_juegos  \
0       Brazil  southamerica     A          1                3   
1       Mexico  northamerica     A          2                3   
2      Croatia        europe     A          3                3   
3     Cameroon        africa     A          4                3   
4  Netherlands        europe     B          1                3   

   juegos_ganados  juegos_empatados  juegos_perdidos  goles_favor  \
0               2                 1                0            7   
1               2                 1                0            4   
2               1                 0                2            6   
3               0                 0                3            1   
4               3                 0                0           10   

   goles_contra goles_diferencia  puntos  clasificado  
0             2                5       7            1  
1             1                3       7            1  
2             6                0       3            0  
3             9               −8       0            0  
4             3                7       9            1

In [61]:
media_goles_contra = media(df["goles_contra"])
varianza_goles_contra = varianza(df["goles_contra"])
print(f"La media de goles a favor es de {media_goles_contra} y su varianza es de {varianza_goles_contra}")

La media de goles a favor es de 4.25 y su varianza es de 4.8125


In [62]:
media_puntos = media(df["puntos"])
varianza_puntos = varianza(df["puntos"])
print(f"La media de goles en contra es de {media_puntos} y su varianza es de {varianza_puntos}")

La media de goles en contra es de 4.21875 y su varianza es de 8.0458984375


In [63]:
media_goles_contra = media(df["goles_contra"])
varianza_goles_contra = varianza(df["goles_contra"])
print(f"La media de puntos obtenidos es de {media_goles_contra} y su varianza es de {varianza_goles_contra}")

La media de puntos obtenidos es de 4.25 y su varianza es de 4.8125


## Desafío 2: Uso de groupby y agg (2 puntos)
Utilizando el método `groupby` de la clase `DataFrame`, en conjunto con la función `.agg` de pandas, calcule la media, la varianza y desviación estándar de la cantidad de goles a favor, en contra y de la cantidad de puntos por continente.

In [3]:
# Utiliza esta celda para aplicar groupby y obtener las métricas solicitadas con agg

In [64]:
df.groupby('continent')[['goles_favor', 'goles_contra', 'puntos']].agg(['mean','var', 'std'])

goles_favor                     goles_contra             \
                    mean       var       std         mean        var   
continent                                                              
africa          3.600000  3.300000  1.816590     5.600000   4.800000   
asia            2.250000  0.916667  0.957427     6.250000   4.250000   
europe          4.769231  6.858974  2.618964     4.000000   3.833333   
northamerica    3.250000  2.250000  1.500000     3.500000  11.000000   
southamerica    5.666667  4.666667  2.160247     2.833333   0.566667   

                          puntos                       
                   std      mean        var       std  
continent                                              
africa        2.190890  2.400000   3.300000  1.816590  
asia          2.061553  0.750000   0.250000  0.500000  
europe        1.957890  4.692308   6.897436  2.626297  
northamerica  3.316625  4.500000  11.000000  3.316625  
southamerica  0.752773  6.833333   3.766667  1.940790

mean       var       std
continent                                 
africa        3.600000  3.300000  1.816590
asia          2.250000  0.916667  0.957427
europe        4.769231  6.858974  2.618964
northamerica  3.250000  2.250000  1.500000
southamerica  5.666667  4.666667  2.160247

**¿En qué continente se observa una mayor cantidad de goles a favor?**

**Respuesta**: En southamerica se observa una mayor cantidad de goles a favor promedio con 5.66

**¿En qué continente se observa una mayor cantidad de goles en contra?**

**Respuesta:** En asia se observa una mayor cantidad de goles en contra promedio con 6.25

**¿En qué continente se observa una mayor cantidad de puntos en promedio?**

**Respuesta**: En southamerica se observa una mayor cantidad de puntos promedio con 6.83

## Desafío 3: Simulaciones (3 puntos)

- Genere una función `generate_pet` que devuelva de forma aleatoria un string `'perro'` o `'gato'` un número `n` de veces. Ejecútela un par de veces.
    - Tip: Puede utilizar la función `np.random.choice` para retornar elementos al azar.
- Aplique la función generate_pet para generar 20 muestras.

In [143]:
# Use este bloque para definir la función generate_pet


def generate_pet2():
    pets = ["perro", "gato"]
    tmp = np.random.choice(pets)
    return tmp



In [144]:
# Use este bloque para generar las 20 muestras usando la función

lista_animales = []

i = 0

while i < 20:
    tmp = generate_pet2()
    lista_animales.append(tmp)
    i += 1
    
perro = 0

probabilidad_perro = lista_animales.count("perro")/len(lista_animales)

print(lista_animales)
probabilidad_perro

['gato', 'gato', 'perro', 'perro', 'gato', 'gato', 'gato', 'gato', 'gato', 'perro', 'perro', 'gato', 'perro', 'gato', 'gato', 'perro', 'perro', 'gato', 'perro', 'perro']


0.45

**¿Cuál es la probabilidad de elegir un perro al azar? ¿Y un gato?**

**Respuesta:** La probabilidad de elegir un perro al azar es de 0.45. Cabe destacar que probablemente si se ejecuta nuevamente el codigo, la probabilidad puede variar ya que es random.

**Agregue `np.random.seed(2)` al inicio del chunk. ¿Qué diferencia hay cuando se ejecuta la función varias veces luego de fijar la semilla?**

**Respuesta**:

## Desafío 4: Función simuladora (4 puntos)
- Genere una función llamada `simulate_pets_prob` que tome como argumento un número finito de simulaciones a generar.
- La función debe simular dos situaciones `'young_pet'` y `'old_pet'`, y contar la ocurrencia de los siguientes casos:
    - De las dos mascotas simuladas (`'young_pet'` y `'old_pet'`), contar las ocasiones donde _por lo menos una_ de las mascotas sea un perro.
    - De las dos mascotas simuladas, contar las ocasiones donde `'old_pet'` sea un perro.
    - De las dos mascotas simuladas, contar las ocasiones donde los dos sean perros.
- El método debe tener una semilla pseudoaleatoria de 1.
- El output de la función debe entregarse en términos de Probabilidad.

In [145]:
def generate_pet(n): 
    return np.random.choice(['Gato', 'Perro'],n)

In [147]:
# Use este bloque para definir la función

def simulate_pets_prob(n):
    np.random.seed(1) 
    cont_1 = 0 
    cont_2 = 0 
    cont_3 = 0
    
    young_pet = generate_pet(n) 
    old_pet = generate_pet(n)
    
    for yp,op in zip(young_pet, old_pet):
        
        if yp == 'Perro'or op == 'Perro': 
            cont_1 += 1 
            
        if op == 'Perro': 
            cont_2 += 1 
            
        if yp == 'Perro'and op == 'Perro': 
            cont_3 += 1
            
    print('Cualquiera sea Perro:', cont_1 / n) # aqui esta dando el %, ya que dicide todos los posibles perros / n (numero entregado)
    print('Old Pet sea Perro:',cont_2 / n) 
    print('Ambos Perro:', cont_3 / n)
    

In [148]:
# Utilice este bloque para realizar el llamado
simulate_pets_prob(1000)

Cualquiera sea Perro: 0.73
Old Pet sea Perro: 0.497
Ambos Perro: 0.245


**De los tres escenarios, ¿Cuál es el menos probable? ¿Cuál es el más probable? ¿Por qué?**

**Respuesta:** El escenario menos probanle es de ambos perros. Por otro lado el escenario mas probable es que
caulquiera sera perro.